File: web/src/types/policy.ts
export type Anchor = 'percent_of_initial' | 'percent_of_current' | 'fixed_usd'
export type Payout = 'monthly' | 'quarterly' | 'yearly'
export type Review = 'once' | 'yearly' | 'monthly'
export type Indicator = 'power_law_quantile' | 'mayer_multiple'
export type Onboarding = 'immediate' | 'deferred_to_euphoria'
export type PresetId =
  | 'classic_fire'
  | 'fixed_pct'
  | 'guardrails'
  | 'valuation_based'
  | 'custom'
export type Phase = 'bear' | 'fair' | 'euphoria'

export interface GuardrailsKnobs {
  enabled: boolean
  ceilingPct: number
  floorPct: number
  adjustPct: number
  prosperity: boolean
}

export interface BufferKnobs {
  enabled: boolean
  years: number
}

export interface ValuationKnobs {
  enabled: boolean
  indicator: Indicator
  fairLow: number
  fairHigh: number
  bearSurplusPct: number
  fairSurplusPct: number
  euphoriaSurplusPct: number
  bufferTargetLowYears: number
  bufferTargetHighYears: number
  safetyValve: number
  onboarding: Onboarding
}

export interface WithdrawalPolicy {
  preset: PresetId
  anchor: Anchor
  ratePct: number
  spendUsd: number
  payout: Payout
  review: Review
  guardrails: GuardrailsKnobs
  buffer: BufferKnobs
  valuation: ValuationKnobs
}

export interface YearResult {
  year: number
  btc: number
  cashUsd: number
  bufferYears: number
  spendUsd: number
  soldBtc: number
  phase: Phase | null
}

export interface PolicyKnobBounds {
  min: number
  max: number
  step: number
  default: number
}

export type PolicyKnobKey =
  | 'ratePct'
  | 'spendUsd'
  | 'ceilingPct'
  | 'floorPct'
  | 'adjustPct'
  | 'bufferYears'
  | 'fairLow'
  | 'fairHigh'
  | 'bearSurplusPct'
  | 'fairSurplusPct'
  | 'euphoriaSurplusPct'
  | 'bufferTargetLowYears'
  | 'bufferTargetHighYears'
  | 'safetyValve'

export const POLICY_BOUNDS: Record<PolicyKnobKey, PolicyKnobBounds> = {
  ratePct: { min: 0, max: 20, step: 0.25, default: 4 },
  spendUsd: { min: 0, max: 10000000, step: 1000, default: 50000 },
  ceilingPct: { min: 0, max: 100, step: 1, default: 20 },
  floorPct: { min: 0, max: 100, step: 1, default: 20 },
  adjustPct: { min: 1, max: 50, step: 1, default: 10 },
  bufferYears: { min: 0, max: 10, step: 0.5, default: 3 },
  fairLow: { min: 0, max: 100, step: 1, default: 50 },
  fairHigh: { min: 0, max: 100, step: 1, default: 85 },
  bearSurplusPct: { min: 0, max: 100, step: 1, default: 0 },
  fairSurplusPct: { min: 0, max: 100, step: 1, default: 0 },
  euphoriaSurplusPct: { min: 0, max: 100, step: 1, default: 8 },
  bufferTargetLowYears: { min: 0, max: 10, step: 0.5, default: 2 },
  bufferTargetHighYears: { min: 0, max: 10, step: 0.5, default: 4 },
  safetyValve: { min: 0, max: 100, step: 1, default: 50 },
}

export const PRESET_LABELS: Record<PresetId, string> = {
  classic_fire: 'Classic FIRE',
  fixed_pct: 'Fixed %',
  guardrails: 'Guardrails',
  valuation_based: 'Valuation-based',
  custom: 'Custom',
}

export const PRESET_DESCRIPTIONS: Record<PresetId, string> = {
  classic_fire: '4% of the initial stack, set once',
  fixed_pct: '4% of the current stack each year',
  guardrails: 'Guyton-Klinger guardrails with prosperity rule',
  valuation_based: 'Phase-aware selling with a cash buffer',
  custom: 'Tune every knob yourself',
}

function baseGuardrails(): GuardrailsKnobs {
  return {
    enabled: false,
    ceilingPct: POLICY_BOUNDS.ceilingPct.default,
    floorPct: POLICY_BOUNDS.floorPct.default,
    adjustPct: POLICY_BOUNDS.adjustPct.default,
    prosperity: true,
  }
}

function baseBuffer(): BufferKnobs {
  return {
    enabled: false,
    years: POLICY_BOUNDS.bufferYears.default,
  }
}

function baseValuation(): ValuationKnobs {
  return {
    enabled: false,
    indicator: 'power_law_quantile',
    fairLow: POLICY_BOUNDS.fairLow.default,
    fairHigh: POLICY_BOUNDS.fairHigh.default,
    bearSurplusPct: POLICY_BOUNDS.bearSurplusPct.default,
    fairSurplusPct: POLICY_BOUNDS.fairSurplusPct.default,
    euphoriaSurplusPct: POLICY_BOUNDS.euphoriaSurplusPct.default,
    bufferTargetLowYears: POLICY_BOUNDS.bufferTargetLowYears.default,
    bufferTargetHighYears: POLICY_BOUNDS.bufferTargetHighYears.default,
    safetyValve: POLICY_BOUNDS.safetyValve.default,
    onboarding: 'deferred_to_euphoria',
  }
}

export function defaultPolicy(): WithdrawalPolicy {
  return policyForPreset('classic_fire')
}

export function policyForPreset(preset: PresetId): WithdrawalPolicy {
  const common = {
    preset,
    anchor: 'percent_of_initial' as Anchor,
    ratePct: POLICY_BOUNDS.ratePct.default,
    spendUsd: POLICY_BOUNDS.spendUsd.default,
    payout: 'yearly' as Payout,
    review: 'once' as Review,
    guardrails: baseGuardrails(),
    buffer: baseBuffer(),
    valuation: baseValuation(),
  }

  switch (preset) {
    case 'classic_fire':
      return common
    case 'fixed_pct':
      return { ...common, anchor: 'percent_of_current', review: 'yearly' }
    case 'guardrails':
      return {
        ...common,
        review: 'yearly',
        guardrails: { ...baseGuardrails(), enabled: true },
      }
    case 'valuation_based':
      return {
        ...common,
        anchor: 'fixed_usd',
        payout: 'monthly',
        review: 'monthly',
        buffer: {
          enabled: true,
          years: POLICY_BOUNDS.bufferTargetHighYears.default,
        },
        valuation: { ...baseValuation(), enabled: true },
      }
    case 'custom':
      return { ...common, review: 'yearly' }
  }
}

export function clampKnobValue(key: PolicyKnobKey, value: number): number {
  const { min, max, default: fallback } = POLICY_BOUNDS[key]
  if (!Number.isFinite(value)) return fallback
  return Math.min(max, Math.max(min, value))
}

export function sanitizePolicy(input: unknown): WithdrawalPolicy {
  const result = defaultPolicy()
  if (typeof input !== 'object' || input === null) return result

  const candidate = input as Record<string, unknown>
  if (candidate.preset === undefined) return result

  const preset = candidate.preset
  if (typeof preset !== 'string' || !(preset in PRESET_LABELS)) {
    return result
  }
  const base = policyForPreset(preset as PresetId)

  const anchor = candidate.anchor
  if (anchor === 'percent_of_initial' || anchor === 'percent_of_current' || anchor === 'fixed_usd') {
    base.anchor = anchor
  }

  const num = (raw: unknown): number | null =>
    typeof raw === 'number' && Number.isFinite(raw) ? raw : null

  const ratePct = num(candidate.ratePct)
  if (ratePct !== null) base.ratePct = clampKnobValue('ratePct', ratePct)
  const spendUsd = num(candidate.spendUsd)
  if (spendUsd !== null) base.spendUsd = clampKnobValue('spendUsd', spendUsd)

  const payout = candidate.payout
  if (payout === 'monthly' || payout === 'quarterly' || payout === 'yearly') {
    base.payout = payout
  }
  const review = candidate.review
  if (review === 'once' || review === 'yearly' || review === 'monthly') {
    base.review = review
  }

  const enumLike = (raw: unknown): string | null =>
    typeof raw === 'string' ? raw : null
  const toggle = (raw: unknown, fallback: boolean): boolean =>
    typeof raw === 'boolean' ? raw : fallback

  if (typeof candidate.guardrails === 'object' && candidate.guardrails !== null) {
    const g = candidate.guardrails as Record<string, unknown>
    base.guardrails.enabled = toggle(g.enabled, base.guardrails.enabled)
    const ceilingPct = num(g.ceilingPct)
    if (ceilingPct !== null) base.guardrails.ceilingPct = clampKnobValue('ceilingPct', ceilingPct)
    const floorPct = num(g.floorPct)
    if (floorPct !== null) base.guardrails.floorPct = clampKnobValue('floorPct', floorPct)
    const adjustPct = num(g.adjustPct)
    if (adjustPct !== null) base.guardrails.adjustPct = clampKnobValue('adjustPct', adjustPct)
    base.guardrails.prosperity = toggle(g.prosperity, base.guardrails.prosperity)
  }

  if (typeof candidate.buffer === 'object' && candidate.buffer !== null) {
    const b = candidate.buffer as Record<string, unknown>
    base.buffer.enabled = toggle(b.enabled, base.buffer.enabled)
    const years = num(b.years)
    if (years !== null) base.buffer.years = clampKnobValue('bufferYears', years)
  }

  if (typeof candidate.valuation === 'object' && candidate.valuation !== null) {
    const v = candidate.valuation as Record<string, unknown>
    base.valuation.enabled = toggle(v.enabled, base.valuation.enabled)
    const indicator = enumLike(v.indicator)
    if (indicator === 'power_law_quantile' || indicator === 'mayer_multiple') {
      base.valuation.indicator = indicator
    }
    const fairLow = num(v.fairLow)
    if (fairLow !== null) base.valuation.fairLow = clampKnobValue('fairLow', fairLow)
    const fairHigh = num(v.fairHigh)
    if (fairHigh !== null) base.valuation.fairHigh = clampKnobValue('fairHigh', fairHigh)
    const bearSurplusPct = num(v.bearSurplusPct)
    if (bearSurplusPct !== null) {
      base.valuation.bearSurplusPct = clampKnobValue('bearSurplusPct', bearSurplusPct)
    }
    const fairSurplusPct = num(v.fairSurplusPct)
    if (fairSurplusPct !== null) {
      base.valuation.fairSurplusPct = clampKnobValue('fairSurplusPct', fairSurplusPct)
    }
    const euphoriaSurplusPct = num(v.euphoriaSurplusPct)
    if (euphoriaSurplusPct !== null) {
      base.valuation.euphoriaSurplusPct = clampKnobValue('euphoriaSurplusPct', euphoriaSurplusPct)
    }
    const bufferTargetLowYears = num(v.bufferTargetLowYears)
    if (bufferTargetLowYears !== null) {
      base.valuation.bufferTargetLowYears = clampKnobValue(
        'bufferTargetLowYears',
        bufferTargetLowYears,
      )
    }
    const bufferTargetHighYears = num(v.bufferTargetHighYears)
    if (bufferTargetHighYears !== null) {
      base.valuation.bufferTargetHighYears = clampKnobValue(
        'bufferTargetHighYears',
        bufferTargetHighYears,
      )
    }
    const safetyValve = num(v.safetyValve)
    if (safetyValve !== null) base.valuation.safetyValve = clampKnobValue('safetyValve', safetyValve)
    const onboarding = enumLike(v.onboarding)
    if (onboarding === 'immediate' || onboarding === 'deferred_to_euphoria') {
      base.valuation.onboarding = onboarding
    }
  }

  return base
}

/** True when the policy's knobs differ from its preset's defaults. */
export function isPolicyDirty(policy: WithdrawalPolicy): boolean {
  return JSON.stringify(policy) !== JSON.stringify(policyForPreset(policy.preset))
}

File: wasm/src/strategies/policy.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Anchor {
    PercentOfInitial,
    PercentOfCurrent,
    FixedUsd,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Payout {
    Monthly,
    Quarterly,
    Yearly,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Review {
    Once,
    Yearly,
    Monthly,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Indicator {
    PowerLawQuantile,
    MayerMultiple,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Onboarding {
    Immediate,
    DeferredToEuphoria,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PresetId {
    ClassicFire,
    FixedPct,
    Guardrails,
    ValuationBased,
    Custom,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Phase {
    Bear,
    Fair,
    Euphoria,
}

pub const RATE_PCT_MIN: f64 = 0.0;
pub const RATE_PCT_MAX: f64 = 20.0;
pub const RATE_PCT_DEFAULT: f64 = 4.0;

pub const SPEND_USD_MIN: f64 = 0.0;
pub const SPEND_USD_MAX: f64 = 10_000_000.0;
pub const SPEND_USD_DEFAULT: f64 = 50_000.0;

pub const GUARDRAIL_THRESHOLD_MIN: f64 = 0.0;
pub const GUARDRAIL_THRESHOLD_MAX: f64 = 100.0;
pub const CEILING_PCT_DEFAULT: f64 = 20.0;
pub const FLOOR_PCT_DEFAULT: f64 = 20.0;

pub const ADJUST_PCT_MIN: f64 = 1.0;
pub const ADJUST_PCT_MAX: f64 = 50.0;
pub const ADJUST_PCT_DEFAULT: f64 = 10.0;

pub const BUFFER_YEARS_MIN: f64 = 0.0;
pub const BUFFER_YEARS_MAX: f64 = 10.0;
pub const BUFFER_YEARS_DEFAULT: f64 = 3.0;

pub const SURPLUS_PCT_MIN: f64 = 0.0;
pub const SURPLUS_PCT_MAX: f64 = 100.0;
pub const EUPHORIA_SURPLUS_PCT_DEFAULT: f64 = 8.0;

pub const FAIR_LOW_MIN: f64 = 0.0;
pub const FAIR_LOW_MAX: f64 = 100.0;
pub const FAIR_LOW_DEFAULT: f64 = 50.0;

pub const FAIR_HIGH_MIN: f64 = 0.0;
pub const FAIR_HIGH_MAX: f64 = 100.0;
pub const FAIR_HIGH_DEFAULT: f64 = 85.0;

pub const BUFFER_TARGET_LOW_MIN: f64 = 0.0;
pub const BUFFER_TARGET_LOW_MAX: f64 = 10.0;
pub const BUFFER_TARGET_LOW_DEFAULT: f64 = 2.0;

pub const BUFFER_TARGET_HIGH_MIN: f64 = 0.0;
pub const BUFFER_TARGET_HIGH_MAX: f64 = 10.0;
pub const BUFFER_TARGET_HIGH_DEFAULT: f64 = 4.0;

pub const SAFETY_VALVE_MIN: f64 = 0.0;
pub const SAFETY_VALVE_MAX: f64 = 100.0;
pub const SAFETY_VALVE_DEFAULT: f64 = 50.0;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Guardrails {
    pub enabled: bool,
    pub ceiling_pct: f64,
    pub floor_pct: f64,
    pub adjust_pct: f64,
    pub prosperity: bool,
}

impl Default for Guardrails {
    fn default() -> Self {
        Guardrails {
            enabled: false,
            ceiling_pct: CEILING_PCT_DEFAULT,
            floor_pct: FLOOR_PCT_DEFAULT,
            adjust_pct: ADJUST_PCT_DEFAULT,
            prosperity: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Buffer {
    pub enabled: bool,
    pub years: f64,
}

impl Default for Buffer {
    fn default() -> Self {
        Buffer {
            enabled: false,
            years: BUFFER_YEARS_DEFAULT,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Valuation {
    pub enabled: bool,
    pub indicator: Indicator,
    pub fair_low: f64,
    pub fair_high: f64,
    pub bear_surplus_pct: f64,
    pub fair_surplus_pct: f64,
    pub euphoria_surplus_pct: f64,
    pub buffer_target_low_years: f64,
    pub buffer_target_high_years: f64,
    pub safety_valve: f64,
    pub onboarding: Onboarding,
}

impl Default for Valuation {
    fn default() -> Self {
        Valuation {
            enabled: false,
            indicator: Indicator::PowerLawQuantile,
            fair_low: FAIR_LOW_DEFAULT,
            fair_high: FAIR_HIGH_DEFAULT,
            bear_surplus_pct: 0.0,
            fair_surplus_pct: 0.0,
            euphoria_surplus_pct: EUPHORIA_SURPLUS_PCT_DEFAULT,
            buffer_target_low_years: BUFFER_TARGET_LOW_DEFAULT,
            buffer_target_high_years: BUFFER_TARGET_HIGH_DEFAULT,
            safety_valve: SAFETY_VALVE_DEFAULT,
            onboarding: Onboarding::DeferredToEuphoria,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WithdrawalPolicy {
    pub preset: PresetId,
    pub anchor: Anchor,
    pub rate_pct: f64,
    pub spend_usd: f64,
    pub payout: Payout,
    pub review: Review,
    pub guardrails: Guardrails,
    pub buffer: Buffer,
    pub valuation: Valuation,
}

fn clamp(v: f64, min: f64, max: f64) -> f64 {
    if !v.is_finite() {
        return min;
    }
    v.min(max).max(min)
}

impl WithdrawalPolicy {
    pub fn classic_fire() -> Self {
        WithdrawalPolicy {
            preset: PresetId::ClassicFire,
            anchor: Anchor::PercentOfInitial,
            rate_pct: RATE_PCT_DEFAULT,
            spend_usd: SPEND_USD_DEFAULT,
            payout: Payout::Yearly,
            review: Review::Once,
            guardrails: Guardrails::default(),
            buffer: Buffer::default(),
            valuation: Valuation::default(),
        }
    }

    pub fn fixed_pct() -> Self {
        WithdrawalPolicy {
            preset: PresetId::FixedPct,
            anchor: Anchor::PercentOfCurrent,
            rate_pct: RATE_PCT_DEFAULT,
            spend_usd: SPEND_USD_DEFAULT,
            payout: Payout::Yearly,
            review: Review::Yearly,
            guardrails: Guardrails::default(),
            buffer: Buffer::default(),
            valuation: Valuation::default(),
        }
    }

    pub fn guardrails() -> Self {
        WithdrawalPolicy {
            preset: PresetId::Guardrails,
            anchor: Anchor::PercentOfInitial,
            rate_pct: RATE_PCT_DEFAULT,
            spend_usd: SPEND_USD_DEFAULT,
            payout: Payout::Yearly,
            review: Review::Yearly,
            guardrails: Guardrails {
                enabled: true,
                ceiling_pct: CEILING_PCT_DEFAULT,
                floor_pct: FLOOR_PCT_DEFAULT,
                adjust_pct: ADJUST_PCT_DEFAULT,
                prosperity: true,
            },
            buffer: Buffer::default(),
            valuation: Valuation::default(),
        }
    }

    pub fn valuation_based() -> Self {
        WithdrawalPolicy {
            preset: PresetId::ValuationBased,
            anchor: Anchor::FixedUsd,
            rate_pct: RATE_PCT_DEFAULT,
            spend_usd: SPEND_USD_DEFAULT,
            payout: Payout::Monthly,
            review: Review::Monthly,
            guardrails: Guardrails::default(),
            buffer: Buffer {
                enabled: true,
                years: BUFFER_TARGET_HIGH_DEFAULT,
            },
            valuation: Valuation {
                enabled: true,
                indicator: Indicator::PowerLawQuantile,
                fair_low: FAIR_LOW_DEFAULT,
                fair_high: FAIR_HIGH_DEFAULT,
                bear_surplus_pct: 0.0,
                fair_surplus_pct: 0.0,
                euphoria_surplus_pct: EUPHORIA_SURPLUS_PCT_DEFAULT,
                buffer_target_low_years: BUFFER_TARGET_LOW_DEFAULT,
                buffer_target_high_years: BUFFER_TARGET_HIGH_DEFAULT,
                safety_valve: SAFETY_VALVE_DEFAULT,
                onboarding: Onboarding::DeferredToEuphoria,
            },
        }
    }

    pub fn custom() -> Self {
        let mut policy = WithdrawalPolicy::classic_fire();
        policy.preset = PresetId::Custom;
        policy.review = Review::Yearly;
        policy
    }

    pub fn for_preset(preset: PresetId) -> Self {
        match preset {
            PresetId::ClassicFire => WithdrawalPolicy::classic_fire(),
            PresetId::FixedPct => WithdrawalPolicy::fixed_pct(),
            PresetId::Guardrails => WithdrawalPolicy::guardrails(),
            PresetId::ValuationBased => WithdrawalPolicy::valuation_based(),
            PresetId::Custom => WithdrawalPolicy::custom(),
        }
    }

    pub fn clamp(&mut self) {
        self.rate_pct = clamp(self.rate_pct, RATE_PCT_MIN, RATE_PCT_MAX);
        self.spend_usd = clamp(self.spend_usd, SPEND_USD_MIN, SPEND_USD_MAX);
        self.guardrails.ceiling_pct = clamp(
            self.guardrails.ceiling_pct,
            GUARDRAIL_THRESHOLD_MIN,
            GUARDRAIL_THRESHOLD_MAX,
        );
        self.guardrails.floor_pct = clamp(
            self.guardrails.floor_pct,
            GUARDRAIL_THRESHOLD_MIN,
            GUARDRAIL_THRESHOLD_MAX,
        );
        self.guardrails.adjust_pct = clamp(
            self.guardrails.adjust_pct,
            ADJUST_PCT_MIN,
            ADJUST_PCT_MAX,
        );
        self.buffer.years = clamp(self.buffer.years, BUFFER_YEARS_MIN, BUFFER_YEARS_MAX);
        self.valuation.fair_low = clamp(self.valuation.fair_low, FAIR_LOW_MIN, FAIR_LOW_MAX);
        self.valuation.fair_high = clamp(self.valuation.fair_high, FAIR_HIGH_MIN, FAIR_HIGH_MAX);
        self.valuation.bear_surplus_pct = clamp(
            self.valuation.bear_surplus_pct,
            SURPLUS_PCT_MIN,
            SURPLUS_PCT_MAX,
        );
        self.valuation.fair_surplus_pct = clamp(
            self.valuation.fair_surplus_pct,
            SURPLUS_PCT_MIN,
            SURPLUS_PCT_MAX,
        );
        self.valuation.euphoria_surplus_pct = clamp(
            self.valuation.euphoria_surplus_pct,
            SURPLUS_PCT_MIN,
            SURPLUS_PCT_MAX,
        );
        self.valuation.buffer_target_low_years = clamp(
            self.valuation.buffer_target_low_years,
            BUFFER_TARGET_LOW_MIN,
            BUFFER_TARGET_LOW_MAX,
        );
        self.valuation.buffer_target_high_years = clamp(
            self.valuation.buffer_target_high_years,
            BUFFER_TARGET_HIGH_MIN,
            BUFFER_TARGET_HIGH_MAX,
        );
        self.valuation.safety_valve = clamp(
            self.valuation.safety_valve,
            SAFETY_VALVE_MIN,
            SAFETY_VALVE_MAX,
        );
    }

    pub fn clamped(mut self) -> Self {
        self.clamp();
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classic_fire_defaults() {
        let p = WithdrawalPolicy::classic_fire();
        assert_eq!(p.preset, PresetId::ClassicFire);
        assert_eq!(p.anchor, Anchor::PercentOfInitial);
        assert!((p.rate_pct - 4.0).abs() < f64::EPSILON);
        assert_eq!(p.payout, Payout::Yearly);
        assert_eq!(p.review, Review::Once);
        assert!(!p.guardrails.enabled);
        assert!(!p.buffer.enabled);
        assert!(!p.valuation.enabled);
    }

    #[test]
    fn fixed_pct_defaults() {
        let p = WithdrawalPolicy::fixed_pct();
        assert_eq!(p.preset, PresetId::FixedPct);
        assert_eq!(p.anchor, Anchor::PercentOfCurrent);
        assert!((p.rate_pct - 4.0).abs() < f64::EPSILON);
        assert_eq!(p.review, Review::Yearly);
        assert!(!p.guardrails.enabled);
    }

    #[test]
    fn guardrails_defaults() {
        let p = WithdrawalPolicy::guardrails();
        assert_eq!(p.preset, PresetId::Guardrails);
        assert_eq!(p.anchor, Anchor::PercentOfInitial);
        assert_eq!(p.review, Review::Yearly);
        assert!(p.guardrails.enabled);
        assert!((p.guardrails.ceiling_pct - 20.0).abs() < f64::EPSILON);
        assert!((p.guardrails.floor_pct - 20.0).abs() < f64::EPSILON);
        assert!((p.guardrails.adjust_pct - 10.0).abs() < f64::EPSILON);
        assert!(p.guardrails.prosperity);
        assert!(!p.buffer.enabled);
        assert!(!p.valuation.enabled);
    }

    #[test]
    fn valuation_based_defaults() {
        let p = WithdrawalPolicy::valuation_based();
        assert_eq!(p.preset, PresetId::ValuationBased);
        assert_eq!(p.anchor, Anchor::FixedUsd);
        assert_eq!(p.payout, Payout::Monthly);
        assert_eq!(p.review, Review::Monthly);
        assert!(!p.guardrails.enabled);
        assert!(p.buffer.enabled);
        assert!(p.valuation.enabled);
        assert_eq!(p.valuation.indicator, Indicator::PowerLawQuantile);
        assert!((p.valuation.fair_low - 50.0).abs() < f64::EPSILON);
        assert!((p.valuation.fair_high - 85.0).abs() < f64::EPSILON);
        assert!((p.valuation.euphoria_surplus_pct - 8.0).abs() < f64::EPSILON);
        assert!((p.valuation.buffer_target_low_years - 2.0).abs() < f64::EPSILON);
        assert!((p.valuation.buffer_target_high_years - 4.0).abs() < f64::EPSILON);
        assert!((p.valuation.safety_valve - 50.0).abs() < f64::EPSILON);
        assert_eq!(p.valuation.onboarding, Onboarding::DeferredToEuphoria);
    }

    #[test]
    fn custom_matches_classic_fire_except_identity() {
        let custom = WithdrawalPolicy::custom();
        let classic = WithdrawalPolicy::classic_fire();
        assert_eq!(custom.preset, PresetId::Custom);
        assert_eq!(custom.anchor, classic.anchor);
        assert_eq!(custom.rate_pct, classic.rate_pct);
        assert_eq!(custom.payout, classic.payout);
        assert_eq!(custom.review, Review::Yearly);
    }

    #[test]
    fn clamp_bounds_rate_and_thresholds() {
        let mut p = WithdrawalPolicy::guardrails();
        p.rate_pct = 999.0;
        p.guardrails.ceiling_pct = 500.0;
        p.guardrails.floor_pct = -10.0;
        p.guardrails.adjust_pct = 0.5;
        p.buffer.years = 99.0;
        p.valuation.euphoria_surplus_pct = -3.0;
        p.clamp();
        assert!((p.rate_pct - RATE_PCT_MAX).abs() < f64::EPSILON);
        assert!((p.guardrails.ceiling_pct - GUARDRAIL_THRESHOLD_MAX).abs() < f64::EPSILON);
        assert!((p.guardrails.floor_pct - GUARDRAIL_THRESHOLD_MIN).abs() < f64::EPSILON);
        assert!((p.guardrails.adjust_pct - ADJUST_PCT_MIN).abs() < f64::EPSILON);
        assert!((p.buffer.years - BUFFER_YEARS_MAX).abs() < f64::EPSILON);
        assert!((p.valuation.euphoria_surplus_pct - SURPLUS_PCT_MIN).abs() < f64::EPSILON);
    }

    #[test]
    fn clamp_handles_non_finite_values() {
        let mut p = WithdrawalPolicy::classic_fire();
        p.rate_pct = f64::NAN;
        p.spend_usd = f64::INFINITY;
        p.clamp();
        assert!((p.rate_pct - RATE_PCT_MIN).abs() < f64::EPSILON);
        assert!((p.spend_usd - SPEND_USD_MIN).abs() < f64::EPSILON);
    }

    #[test]
    fn serde_roundtrip_preserves_every_preset() {
        let presets = [
            WithdrawalPolicy::classic_fire(),
            WithdrawalPolicy::fixed_pct(),
            WithdrawalPolicy::guardrails(),
            WithdrawalPolicy::valuation_based(),
            WithdrawalPolicy::custom(),
        ];
        for p in presets {
            let json = serde_json::to_string(&p).unwrap();
            let restored: WithdrawalPolicy = serde_json::from_str(&json).unwrap();
            assert_eq!(restored.preset, p.preset);
            assert_eq!(restored.anchor, p.anchor);
            assert!((restored.rate_pct - p.rate_pct).abs() < f64::EPSILON);
            assert!((restored.spend_usd - p.spend_usd).abs() < f64::EPSILON);
            assert_eq!(restored.payout, p.payout);
            assert_eq!(restored.review, p.review);
            assert_eq!(restored.guardrails.enabled, p.guardrails.enabled);
            assert!((restored.guardrails.ceiling_pct - p.guardrails.ceiling_pct).abs() < f64::EPSILON);
            assert!((restored.guardrails.floor_pct - p.guardrails.floor_pct).abs() < f64::EPSILON);
            assert!((restored.guardrails.adjust_pct - p.guardrails.adjust_pct).abs() < f64::EPSILON);
            assert_eq!(restored.guardrails.prosperity, p.guardrails.prosperity);
            assert_eq!(restored.buffer.enabled, p.buffer.enabled);
            assert!((restored.buffer.years - p.buffer.years).abs() < f64::EPSILON);
            assert_eq!(restored.valuation.enabled, p.valuation.enabled);
            assert_eq!(restored.valuation.indicator, p.valuation.indicator);
            assert!((restored.valuation.fair_low - p.valuation.fair_low).abs() < f64::EPSILON);
            assert!((restored.valuation.fair_high - p.valuation.fair_high).abs() < f64::EPSILON);
            assert!((restored.valuation.bear_surplus_pct - p.valuation.bear_surplus_pct).abs() < f64::EPSILON);
            assert!((restored.valuation.fair_surplus_pct - p.valuation.fair_surplus_pct).abs() < f64::EPSILON);
            assert!((restored.valuation.euphoria_surplus_pct - p.valuation.euphoria_surplus_pct).abs() < f64::EPSILON);
            assert!((restored.valuation.buffer_target_low_years - p.valuation.buffer_target_low_years).abs() < f64::EPSILON);
            assert!((restored.valuation.buffer_target_high_years - p.valuation.buffer_target_high_years).abs() < f64::EPSILON);
            assert!((restored.valuation.safety_valve - p.valuation.safety_valve).abs() < f64::EPSILON);
            assert_eq!(restored.valuation.onboarding, p.valuation.onboarding);
        }
    }
}
